common/locks/priority_semaphore.go

```diff
@@ -0,0 +1,33 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package locks
+
+import "context"
+
+type (
+	PrioritySemaphore interface {
+		Acquire(ctx context.Context, priority Priority, n int) error
+		TryAcquire(priority Priority, n int) bool
+		Release(priority Priority, n int)
```

Comment:
If we need the release to have the same priority as the acquire then the acquire should return a "hold" object that does that for you. Otherwise we're inviting some NASTY bugs into our system. 

If the Acquire returns a small struct that contains a reference to both the semaphore and the request's priority, we can defer "hold.Release()" and be done with it

Reply:
It's not in Go but I built this for a hobby project for this exact reason at https://github.com/tdeebswihart/zdb/blob/main/lib/sync/latch.zig

Reply:
Yeah, that is right. release doesn't need priority. Same for TryAcquire(). It gets the lock if there is resource available irrespective of priority.